src/librustc_typeck/lib.rs

```diff
@@ -138,6 +140,30 @@ pub struct TypeAndSubsts<'tcx> {
     ty: Ty<'tcx>,
 }
 
+fn allow_type_alias_enum_variants<'a, 'gcx, 'tcx>(tcx: TyCtxt<'a, 'gcx, 'tcx>,
+                                                  ty_hir: &hir::Ty,
+                                                  span: Span) -> bool {
+    let allow_feature = tcx.features().type_alias_enum_variants;
+    if !allow_feature {
+        // Only print error if we know the type is an alias.
+        if let hir::TyKind::Path(hir::QPath::Resolved(None, ref path)) = ty_hir.node {
+            if let Def::TyAlias(_) = path.def {
+                let mut err = tcx.sess.struct_span_err(
+                    span,
+                    "type alias enum variants are not yet allowed"
```

Comment:
Nit: "not yet allowed" -> "unstable" or something.

Reply:
Do we want to put that message in stable Rust though? It might be misleading.

Reply:
Well, that's what feature errors typically say, worked ok so far.
```rust
macro m() {} // ERROR `macro` is experimental (see issue #39412)
```
Probably not too important, though.

Reply:
Okay. Sorted, anyway. :-)